dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -230,18 +235,31 @@ class NVDECCache {
     }
 
  private:
-    NVDECCache() {}
+    NVDECCache() {
+      int num_devices = 0;
+      CUDA_CALL(cudaGetDeviceCount(&num_devices));
+      dec_cache.resize(num_devices);
+      access_lock.resize(num_devices);
+    }
 
     ~NVDECCache() {
-      std::scoped_lock lock(access_lock);
-      for (auto &it : dec_cache) {
-        cuvidDestroyDecoder(it.second.decoder);
+      for (size_t i = 0; i < dec_cache.size(); ++i) {
+        auto &dec_cache_elm = dec_cache[i];
+        std::scoped_lock lock(access_lock[i]);
+        for (auto &it : dec_cache_elm) {
+          cuvidDestroyDecoder(it.second.decoder);
+        }
       }
     }
-
-    std::unordered_multimap<cudaVideoCodec, DecInstance> dec_cache;
-
-    std::mutex access_lock;
+    using codec_map = std::unordered_multimap<cudaVideoCodec, DecInstance>;
+    std::vector<codec_map> dec_cache;
+
+    struct mutex_wrapper : std::mutex {
```

Comment:
What is this wrapper for?

Reply:
To create a vector of mutexes (raw mutex doesn't allow that). I'm open to other options.

Reply:
The problem is `resize`, but you don't have to use it.
```
mutexes = std::vector<std::mutex>(32);
```

Reply:
Reworked